src/lib/Actor.hs

```diff
@@ -1,178 +1,67 @@
-data BackChan a = BackChan (IORef [a]) (Chan a)
-data ActorConfig m = ActorConfig { selfProc  :: Proc
-                                 , selfChan  :: BackChan (Msg m)
-                                 , linksPtr  :: IORef [Proc] }
-
-newtype Actor m a = Actor (ReaderT (ActorConfig m) IO a)
-  deriving (Functor, Applicative, Monad, MonadIO)
-
-runActor :: Actor msg a -> IO a
-runActor (Actor m) = do
-  linksRef   <- newIORef []
-  chan <- newBackChan
-  tid <- myThreadId
-  let p = Proc Trap tid (asErrPChan chan)
-  runReaderT m (ActorConfig p chan linksRef)
-
-subChan :: (a -> b) -> PChan b -> PChan a
-subChan f chan = PChan (sendPChan chan . f)
-
--- TODO: kill :: Proc -> Actor m ()
-
-spawn :: MonadActor msg m => CanTrap -> Actor msg' () -> m (Proc, PChan msg')
-spawn canTrap body = liftIO $ spawnIO canTrap [] body
-
-spawnLink :: MonadActor msg m => CanTrap -> Actor msg' () -> m (Proc, PChan msg')
-spawnLink canTrap body = do
-  cfg <- actorCfg
-  liftIO $ do
-    links <- readIORef (linksPtr cfg)
-    (child, childChan) <- spawnIO canTrap [selfProc cfg] body
-    -- potential bug if we get killed right here, before we've linked the child.
-    -- 'mask' from Control.Exception might be a solution
-    writeIORef (linksPtr cfg) (child : links)
-    return (child, childChan)
-
-spawnIO :: CanTrap -> [Proc] -> Actor msg () -> IO (Proc, PChan msg)
-spawnIO canTrap links (Actor m) = do
-  linksRef   <- newIORef links
-  chan <- newBackChan
-  tid <- forkIO $ do
-    tid <- myThreadId
-    let self = Proc canTrap tid (asErrPChan chan)
-    runReaderT m (ActorConfig self chan linksRef)
-      `catch` (\e ->
-         do linked <- readIORef linksRef
-            putStrLn $ "Error:\n" ++ show (e::SomeException)
-            mapM_ (cleanup (show (e::SomeException)) self) linked)
-  return (Proc canTrap tid (asErrPChan chan), asPChan chan)
-  where
-   cleanup :: String -> Proc -> Proc -> IO ()
-   cleanup s failed (Proc Trap   _ errChan) = sendFromIO errChan (failed, s)
-   cleanup _ _      (Proc NoTrap linked  _) = void $ forkIO $ killThread linked
 
-asPChan :: BackChan (Msg a) -> PChan a
-asPChan (BackChan _ chan) = PChan (writeChan chan . NormalMsg)
+import Util (onFst, onSnd)
 
-asErrPChan :: BackChan (Msg a) -> PChan (Proc, String)
-asErrPChan (BackChan _ chan) = PChan (writeChan chan . uncurry ErrMsg)
+-- Micro-actors.
 
-send :: MonadActor msg' m => PChan msg -> msg -> m ()
-send p x = liftIO (sendFromIO p x)
+-- In this model, an "actor" is just an IO computation (presumably
+-- running on its own Haskell thread) that receives messages on a
+-- Control.Concurrent.Chan channel.  The idea is that the actor thread
+-- only receives information (or synchronization) from other threads
+-- through messages sent on that one channel, and no other thread
+-- reads messages from that channel.
 
-sendFromIO :: PChan msg -> msg -> IO ()
-sendFromIO = sendPChan
+type Actor a = Chan a -> IO ()
```

Comment:
Can actors send messages back? If not, why do we have a type alias for send-only channels, but not for receive-only channels?

Reply:
An actor will generally not send a message on its own channel, because that would constitute sending a message to itself (though it could do that, to break up a long computation into chunks that can be interspersed with processing other messages).  However, an actor _may_ send a send-only view of its own channel to other actors, to subscribe themselves to notifications of asynchronous events.  So at creation time, an actor is given a two-way channel to allow that to happen.

Code comment added.